### PR TITLE
Integrate UI-as-code features into the language specification

### PR DESCRIPTION
**Pull request by eernstg**
We have a couple of issues with the completeness of the language specification with this addition: It refers to context types and type inference many times, with no definitions.

The current approach is to simply give a brief intuitive explanation when those concepts are first used, such that we just need to adjust references to these concepts when we add suitable sections about inference etc.

Note that I've changed the specification of element evaluation such that they rely on the simplest possible syntax (\FOR{} statements), rather than specifying actions such as evaluating `collection.iterator` etc., because the \FOR{} statements are already specified to do just that.

Similarly, I'm using the notion of `$S$ implements $T$` which means that `$T$` is a superinterface (not just a supertype) of `$S$`, which means that we already know that `$S$` cannot be a bottom/null type when `$T$` is `Iterable` or `Map`. This usage of the word `implements` is consistent with everything in the spec already, but it hasn't been defined explicitly before now.

Also, I'm using the phrase `the type argument of $S$ at 'Iterable'` which will map the actual type arguments of `$S$` through the suitable superinterface links, substituting type arguments as needed (for instance, the type argument of `List<int>` at `Iterable` is `int`, and with `class C<X> extends B<Map<X, X>> {}`, the type argument of `C<int>` at `B` is `Map<int, int>`). 

I've changed a number of compile-time errors to a big comment at the end of \ref{collectionLiterals}, because they are already implied by other parts of the text. We could mention them in commentary or delete them, whatever is considered more useful for the reader.

### CODE REVIEW

File: specification/dartLangSpec.tex [hunk 1]
```diff
@@ -7020,496 +7064,1831 @@ \subsection{Symbols}
...
 }
 
 \LMHash{}%
-It is a compile-time error if
-either a key or a value of an entry in a constant map literal
+It is a compile-time error if a collection literal element in a constant set literal
 is not a constant expression.
 It is a compile-time error if
-the operator \lit{==} of the key of an entry in a constant map literal
+the operator \lit{==} of an element expression in a constant map literal
```
Comment (munificent) [inline specification/dartLangSpec.tex]: "map" -> "set".

File: specification/dartLangSpec.tex [hunk 2]
```diff
@@ -7020,496 +7064,1831 @@ \subsection{Symbols}
...
+or it may be a computational element specifying
+how to obtain zero or more entities,
+of the form \synt{ifElement} or \synt{forElement}.
+
+\commentary{%
+Terms derived from \synt{element},
+and the ability to build collections from them,
+is also known as
+\Index{UI-as-code}.%
+}
```
[the remark this answers is not in the document]
Comment (eernstg) [reply specification/dartLangSpec.tex]: I was thinking that we've used the phrase UI-as-code so much that it's helpful to be able to search it. Would you prefer that we take it out?

File: specification/dartLangSpec.tex [hunk 3]
```diff
@@ -7020,496 +7064,1831 @@ \subsection{Symbols}
...
+
+\LMHash{}%
+The context type $P$
+(\ref{setAndMapLiteralDisambiguation})
+for each element of \metavar{list} is
+obtained from the context type of \metavar{list}.
+If downwards inference constrains the type of \metavar{list}
+%% TODO(eernst): Confirm that the context type should allow for
+%% `List` as well as `Iterable`. The implementations (unsurprisingly)
+%% do that, but the feature spec only allowed for `Iterable`.
```
[the remark this answers is not in the document]
Comment (eernstg) [reply specification/dartLangSpec.tex]: OK, good! Deleting the TODO.

File: specification/dartLangSpec.tex [hunk 4]
```diff
@@ -7020,496 +7064,1831 @@ \subsection{Symbols}
...
+obtained by evaluating $s_{\metavar{syntax}}$
+is the concatenation of the sequences of objects
+obtained by evaluating each element $\ell_j$, $j \in 1 .. k$:
+$s_{\metavar{object}}=\EvaluateElement{\ell_1}+\ldots+\EvaluateElement{\ell_k}$,
+where \EvaluateElement{\ell_j} denotes the object sequence yielded by
+evaluation of a single collection literal element $\ell_j$.
+
+\LMHash{}%
+When a pseudo-statement of the form
+\code{$s = s + \EvaluateElement{\ell}$;}
```
[the remark this answers is not in the document]
Comment (eernstg) [reply specification/dartLangSpec.tex]: Yes. Done.

File: specification/dartLangSpec.tex (hunk 1 above)
Comment (eernstg) [reply specification/dartLangSpec.tex]: Done.

**Pull request merged**